src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java

```diff
@@ -2130,6 +2131,17 @@ void addSymbol(Symbol sym, LambdaSymbolKind skind) {
                 }
             }
 
+            private boolean outerThisReachable(Type current, Symbol target) {
+                if (current == Type.noType || typesUnderConstruction.contains(current.tsym)) {
+                    return false;
+                } else if (current.tsym.isSubClass(target, types)) {
+                    return true;
+                } else {
+                    return current.tsym.hasOuterInstance() &&
```

Comment:
note that `hasOuterInstance` takes into account the `NOOUTERTHIS` flag, that is, whether the inner class can have an enclosing instance at all.